### PR TITLE
refactor/lockfile

### PR DESCRIPTION
**Pull request by fcanela**
Extracts the Lockfile logic so it is easier to remove once we normalize the repository to be deployable to our infrastructure. Provides tests.

This PR is based in #375, so the diff may contain content from that PR if it has not been merged yet.

---

## PR-Codex overview
This PR focuses on implementing a locking mechanism to prevent multiple instances of a process from running simultaneously. It adds functions for claiming and releasing locks, along with corresponding tests to ensure functionality.

### Detailed summary
- Added `claimLock` and `releaseLock` functions in `lock.ts`.
- Implemented `getLockFilePath` to generate lock file paths.
- Introduced `LockfileExistsError` for error handling.
- Updated `initialize` function in `relayerHelpers.ts` to use locking functions.
- Added tests for locking functionality in `lock.test.ts`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

## Summary by CodeRabbit

- **New Features**
	- Introduced a locking mechanism to manage concurrent processes, ensuring only one instance can run at a time.
	- Added a new Jest configuration file for improved testing capabilities, including coverage reporting.
	- New script added to `package.json` for running tests with coverage.
	- Updated ESLint configuration to recognize Jest global variables.

- **Bug Fixes**
	- Streamlined shutdown management process by removing the previous `ShutdownManager` class and integrating lock management functions.

- **Tests**
	- Added unit tests for the locking mechanism to ensure proper functionality and error handling.

### CODE REVIEW

File: relayer-cli/src/utils/relayerHelpers.ts [hunk 1]
```diff
@@ -1,29 +1,9 @@
...
 
 async function initialize(chainId: number, network: string): Promise<number> {
-  const lockFileName = "./state/" + network + "_" + chainId + ".pid";
-
-  if (fs.existsSync(lockFileName)) {
-    console.log("Skipping chain with process already running, delete pid file to force", chainId);
-    throw new Error("Already running");
-  }
-  fs.writeFileSync(lockFileName, process.pid.toString(), { encoding: "utf8" });
+  claimLock(network, chainId);
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/relayerHelpers.ts:6]: There is no need for that. The LockfileExistsError already has a nice error message that is suitable to be displayed. This suggestion just makes the code more complex and introduces unrequired branching

File: relayer-cli/src/utils/shutdownManager.test.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    it("should return true when isShuttingDown is false", () => {
+      const instance = new ShutdownManager(false);
+      expect(instance.getIsShuttingDown()).toBe(false);
+    });
+  });
+
+  describe("triggerShutdown", () => {
+    it.todo("should set isShuttingDown to true");
+  });
+});
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/shutdownManager.test.ts:36]: Agree, and will be fixed in #375 and ported to this PR

File: relayer-cli/src/utils/lock.test.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      claimLock(network, chainId, deps);
+
+      expect(deps.fileExistsFn).toHaveBeenCalledTimes(1);
+      expect(deps.writeFileFn).toHaveBeenCalledTimes(1);
+
+      const [path, pid] = deps.writeFileFn.mock.calls[0];
+      expect(path).toBe(expectedLockFilePath);
+      expect(pid).toBe(process.pid.toString());
+    });
+  });
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/lock.test.ts:50]: File write errors will throw an exception that will bubble up. Tests are already checking that the PID in the file is the expected one. I am not going to test that the `fs` NodeJS library is operating correctly: that is part of the native library responsabilities/tests.

File: relayer-cli/src/utils/lock.test.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        unlinkFileFn: jest.fn(),
+      };
+
+      releaseLock(network, chainId, deps);
+
+      expect(deps.fileExistsFn).toHaveBeenCalledTimes(1);
+      expect(deps.unlinkFileFn).not.toHaveBeenCalled();
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/lock.test.ts:84]: That case will bubble up and interrupt the application execution

File: relayer-cli/src/utils/lock.test.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
+import { claimLock, getLockFilePath, LockfileExistsError, releaseLock } from "./lock";
+
+describe("Lock", () => {
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/lock.test.ts:3]: Will be fixed in #375 and ported to this PR

File: relayer-cli/src/utils/relayerHelpers.ts [hunk 6]
```diff
@@ -54,10 +34,7 @@ async function updateStateFile(chainId: number, createdTimestamp: number, nonceF
   };
   fs.writeFileSync(chain_state_file, JSON.stringify(json), { encoding: "utf8" });
 
-  const lockFileName = "./state/" + network + "_" + chainId + ".pid";
-  if (fs.existsSync(lockFileName)) {
-    fs.unlinkSync(lockFileName);
-  }
+  releaseLock(network, chainId);
```
Comment (mani99brar) [inline relayer-cli/src/utils/relayerHelpers.ts:37]: I believe we can remove `releaseLock` from here, and use it in `setupExitHandlers` during `cleanup`, [here](https://github.com/kleros/vea/blob/71c528085c53941ea625c995cee39774ce800f9b/relayer-cli/src/utils/relayerHelpers.ts#L43).

File: relayer-cli/package.json [hunk 7]
```diff
@@ -23,6 +24,10 @@
     "web3-batched-send": "^1.0.3"
   },
   "devDependencies": {
-    "ts-node": "^10.9.2"
+    "@types/jest": "^29.5.14",
+    "jest": "^29.7.0",
+    "ts-jest": "^29.2.5",
+    "ts-node": "^10.9.2",
+    "yarn": "^1.22.22"
```
Comment (mani99brar) [inline relayer-cli/package.json:31]: Confused about `yarn` defined again in devDependencies, and with a different version.

Comment (fcanela) [reply relayer-cli/package.json:31]: @mani99brar I do not recall installing it explictly, so not sure why my local yarn decided to do so. I'll remove it. Thanks for the hint!

File: relayer-cli/src/utils/lock.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  dependencies: ClaimLockDependencies = {
+    fileExistsFn: fs.existsSync,
+    writeFileFn: fs.writeFileSync,
+  }
+) {
+  const path = getLockFilePath(network, chain);
+  const { fileExistsFn, writeFileFn } = dependencies;
+
+  if (fileExistsFn(path)) throw new LockfileExistsError(path);
+  writeFileFn(path, process.pid.toString(), { encoding: "utf8" });
```
[the remark this answers is not in the document]
Comment (fcanela) [reply relayer-cli/src/utils/lock.ts:54]: Agree, but the plan is to move away from lockfile as this is going to be deployed as a docker container in a cloud solution. Having a file based implementation is not very friendly with that, so we will move to something that is comfortable to use for both local and cloud deployments.

File: relayer-cli/src/utils/relayerHelpers.ts (hunk 6 above)
Comment (fcanela) [reply relayer-cli/src/utils/relayerHelpers.ts:37]: @coderabbitai can you create an issue for this?

**Review by mani99brar — APPROVED** [review]

**Pull request merged**